File: Leetcode/Leetcode-DSA-course/Trees-and-Graphs/530-minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int getMinimumDifference(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        dfs(root, values);
        int ans = Integer.MAX_VALUE;
        for (int i = 1; i < values.size(); i++) {
            ans = Math.min(ans, values.get(i) - values.get(i - 1));
        }
        return ans;
    }

    void dfs(TreeNode root, List<Integer> values) {
        if (root == null)
            return;
        dfs(root.left, values);
        values.add(root.val);
        dfs(root.right, values);
    }
}

File: Leetcode/Leetcode-DSA-course/Trees-and-Graphs/530-minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        vector<int> values;
        dfs(root, values);
        int ans = INT_MAX;
        for (int i = 1; i < values.size(); i++)
            ans = min(ans, values[i] - values[i - 1]);
        return ans;
    }
    void dfs(TreeNode* root, vector<int>& values) {
        if (root == nullptr)
            return;
        dfs(root->left, values);
        values.push_back(root->val);
        dfs(root->right, values);
    }
};